docs/reference/module-api/module-overview.md

```diff
@@ -908,12 +829,146 @@ module.exports = {
+          await query.finalize();
+
+          const pipeline = [
+            { $match: query.get('criteria') },
+            { $sample: { size: 1 } }
+          ];
+          const result = await self.apos.doc.db.aggregate(pipeline)
+            .toArray();
+
+          return result[0];
+        }
+      }
+    };
+  }
+};
+```
+
 #### `extendQueries(self)`
 
+Extend the behavior of existing event handlers (set in `queries`) in the `extendQueries` section. This function must return an object as described in `queries`.
+
+Each extended query builder or method should accept the original function as `_super` followed by its original arguments. Extended query builders and methods will be matched with the base class builder or method using the same name. Methods should return data in a similar format to the existing API route.
+
 ### `middleware(self)`
 
+Add standard Express middleware to be called on *every* request. The `middleware` function takes the module as an argument and must return an object of [middleware functions](https://expressjs.com/en/guide/using-middleware.html). This is a good place to import third-party middleware.
+
+Note that it is often better to add an event handler or `await` a method in an API route instead.
+
+`checkIp` could also be an object with a `before` property and a `middleware` property, in which case it would run `before` the middleware of the module with the specified name
+
+```javascript
+// modules/limiter/index.js
+module.exports = {
+  // ...
+  middleware(self, options) {
+    return {
+      checkIp(req, res, next) {
+        // Restrict access by IP address, in a crude way.
+        const allowlist = [ '127.0.0.1', '::1' ];
+
+        if (!allowlist.includes(req.connection.remoteAddress)) {
+          return res.status(403).send('forbidden');
+        }
+        return next();
+      }
+    };
+  }
+};
+```
+
+If the middleware function must run before another specific module's middleware, set the returned object key to an object with `before` and `middleware` properties. `before` would be set to the name of the module whose middleware must run after the new function, which is now the value of `middleware`.
+
+```javascript
+// modules/limiter/index.js
+module.exports = {
+  // ...
+  middleware(self, options) {
+    return {
+      checkIp: {
+        // 👇 Same as above, but with `before` and `middleware` properties.
+        before: '@apostrophecms/login',
+        middleware: function (req, res, next) {
+          // Restrict access by IP address, in a crude way.
+          const allowlist = [ '127.0.0.1', '::1' ];
+
+          if (!allowlist.includes(req.connection.remoteAddress)) {
+            return res.status(403).send('forbidden');
+          }
+          return next();
+        }
+      }
+    };
+  }
+};
+```
+
 ### `tasks(self)`
 
-## "Cascading" settings
+`tasks` takes the module as an argument and returns an object of functions that define command line tasks. Task properties include:
```

Comment:
I guess you don't want to say "object of methods" because that is easily confused with apostrophe module methods. OK.

Reply:
some of this is me copying other text lazily.